Here’s an example of my script in which I make multilayer perceptron on some financial data. The data has 53 independent variables (X) and one independent variable (y). It is time series data and X exhibits plenty of time-series related information. The goal is to fit MP as good as possible.

In [12]:
### Import packages:
import numpy as np
import sys
import theano
import theano.tensor as T
from keras.models import Sequential
import keras

In [13]:
### Load data and count the number of samples.
input_data_all = np.load('train_sample.npy')
output_data_all= np.load('train_target.npy')
n_samples= input_data_all.shape[0]
print(n_samples)

96674


In [14]:
### shuffle 2 data series together randomly.
def shuffle_together(input_1, input_2):
    if input_1.shape[0]!= input_2.shape[0]:
        print ("Problem, y and x array are not of the same shape.")
        return None
    c= np.arange(input_1.shape[0])
    np.random.shuffle(c)
    return input_1[c], input_2[c], c

In [15]:
### splitting on train and validation input (note that it is splitted randomly).

input_data_all, output_data_all, _ = shuffle_together(input_data_all, output_data_all)
n_validation_samples= int(0.1*n_samples)
validation_input= input_data_all[-1*n_validation_samples:]
validation_output= output_data_all[-1*n_validation_samples:]
n_train_samples= n_samples-n_validation_samples
train_input= input_data_all[:n_train_samples]
train_output= output_data_all[:n_train_samples]
print(train_input.shape)
print(train_output.shape)
### Now we define matrix x and vector y in theano.
x=T.matrix('x', dtype= theano.config.floatX)
y=T.vector('y', dtype= theano.config.floatX)


(87007, 53)
(87007,)


In [16]:
train_input.shape[1]

53

In [17]:
# initialize model
model = keras.models.Sequential()
### First layer.  Linear activation layer.
model.add(keras.layers.Dense(input_data_all.shape[1], input_shape=(train_input.shape[1],),activation='linear',
                bias_initializer=keras.initializers.RandomUniform(minval=-1, maxval=1, seed=None),
                kernel_initializer=keras.initializers.RandomUniform(minval=-1, maxval=1, seed=None)))

# add hidden layer(s) with relu and tanh activation functions.
model.add(keras.layers.Dense(units=40,kernel_initializer=keras.initializers.RandomUniform(minval=-1, maxval=1, seed=None),
        bias_initializer=keras.initializers.RandomUniform(minval=-0.05, maxval=0.05, seed=None),
        activation='relu')
    )          
model.add(keras.layers.Dense(units=25,kernel_initializer=keras.initializers.RandomUniform(minval=-0.05, maxval=0.05, seed=None),
        bias_initializer=keras.initializers.RandomUniform(minval=-0.05, maxval=0.05, seed=None),
        activation='tanh')
    )          
model.add(keras.layers.Dense(units=15,kernel_initializer=keras.initializers.RandomUniform(minval=-0.05, maxval=0.05, seed=None),
        bias_initializer=keras.initializers.RandomUniform(minval=-0.05, maxval=0.05, seed=None),
        activation='tanh')
    )          
model.add(keras.layers.Dense(units=8,kernel_initializer=keras.initializers.RandomUniform(minval=-0.05, maxval=0.05, seed=None),
        bias_initializer=keras.initializers.RandomUniform(minval=-0.05, maxval=0.05, seed=None),
        activation='tanh')
    )          
model.add(keras.layers.Dense(units=4,kernel_initializer=keras.initializers.RandomUniform(minval=-0.05, maxval=0.05, seed=None),
        bias_initializer=keras.initializers.RandomUniform(minval=-1, maxval=1, seed=None),
        activation='relu')
    )          
### Final layer is again linear.
model.add(keras.layers.Dense(units=1,kernel_initializer=keras.initializers.RandomUniform(minval=-0.05, maxval=0.05, seed=None),
        bias_initializer=keras.initializers.RandomUniform(minval=-0.05, maxval=0.05, seed=None),
        activation='linear')
    )          


In [18]:
# define SGD optimizer
sgd_optimizer = keras.optimizers.Adam(learning_rate=0.001, beta_1=0.9, beta_2=0.999, epsilon=1e-012)

# compile model
model.compile(
    optimizer=sgd_optimizer,
    loss='mean_squared_error'
)

In [19]:
# train model
history = model.fit(
    input_data_all, output_data_all,
    batch_size=300, epochs=100000,
    verbose=1, validation_split=0.1,validation_freq=100
)

Train on 87006 samples, validate on 9668 samples
Epoch 1/100000
87006/87006 [==============================] - 9s 103us/step - loss: 1.2347
Epoch 2/100000
87006/87006 [==============================] - 12s 141us/step - loss: 0.0443ETA: 2s - los - ETA: 2s  - ETA: 1s  - ETA: 0s -
Epoch 3/100000
87006/87006 [==============================] - 14s 163us/step - loss: 0.04173s  - ETA:
Epoch 4/100000
87006/87006 [==============================] - 7s 79us/step - loss: 0.0417
Epoch 5/100000
87006/87006 [==============================] - 4s 45us/step - loss: 0.0417
Epoch 6/100000
87006/87006 [==============================] - 6s 74us/step - loss: 0.0417: 1s - lo
Epoch 7/100000
87006/87006 [==============================] - 8s 87us/step - loss: 0.0417
Epoch 8/100000
59100/87006 [===================>..........] - ETA: 2s - loss: 0.0418- ETA:  - E - ETA: 2s - loss: 0.04

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.100987). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)
C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.120425). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


60000/87006 [===================>..........] - ETA: 2s - loss: 0.0419

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.130463). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


61200/87006 [====================>.........] - ETA: 3s - loss: 0.0419

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.109220). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 16s 184us/step - loss: 0.0417
Epoch 9/100000
87006/87006 [==============================] - 19s 218us/step - loss: 0.041320s - loss: 0. - ETA: 20 - ETA: 17s  - ETA: - ETA: 7s -  - ETA: 6s - loss: 0 - ETA:   - ETA: 0s - loss: 0
Epoch 10/100000
87006/87006 [==============================] - 6s 69us/step - loss: 0.0391
Epoch 11/100000
87006/87006 [==============================] - 20s 230us/step - loss: 0.0385
Epoch 12/100000
87006/87006 [==============================] - 24s 277us/step - loss: 0.0379A: 1s - loss: 0 - ETA: 1s - loss: 
Epoch 13/100000
  900/87006 [..............................] - ETA: 26s - loss: 0.0381

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.131560). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


 1800/87006 [..............................] - ETA: 35s - loss: 0.0378

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.104546). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)
C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.118551). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 14s 163us/step - loss: 0.0375
Epoch 14/100000
87006/87006 [==============================] - 9s 104us/step - loss: 0.0372 1s
Epoch 15/100000
87006/87006 [==============================] - 7s 78us/step - loss: 0.0373
Epoch 16/100000
87006/87006 [==============================] - 12s 140us/step - loss: 0.0362
Epoch 17/100000
87006/87006 [==============================] - 11s 121us/step - loss: 0.0352 ETA: 9s  - ETA: 
Epoch 18/100000
87006/87006 [==============================] - 15s 169us/step - loss: 0.03013s -  - ETA: 2s - loss:  - ETA
Epoch 19/100000
87006/87006 [==============================] - 14s 156us/step - loss: 0.0170TA: 8s - loss: 0.025 - ETA: 8s - loss: - ET
Epoch 20/100000
87006/87006 [==============================] - 11s 128us/step - loss: 0.00420s - loss: 0.0
Epoch 21/100000
87006/87006 [==============================] - 12s 133us/step - loss: 0.0027
Epoch 22/100000
87006/87006 [==============================] - 13s 154us/st

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.106291). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 14s 165us/step - loss: 0.00135s -  - ETA: 4s - l - ETA: 4 - ETA: 3s  - ETA: 1s - loss: 0 - ETA: 
Epoch 49/100000
  900/87006 [..............................] - ETA: 23s - loss: 0.0040

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.132552). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 6s 72us/step - loss: 0.0013
Epoch 50/100000
87006/87006 [==============================] - 7s 85us/step - loss: 0.0011 ETA: 0s - loss: 0.00
Epoch 51/100000
87006/87006 [==============================] - 6s 64us/step - loss: 9.1541e-04
Epoch 52/100000
87006/87006 [==============================] - 11s 129us/step - loss: 0.00303s - loss - ETA: 3s - l - ETA: 3s - loss: 0 - ETA: 2s - l
Epoch 53/100000
  600/87006 [..............................] - ETA: 21s - loss: 0.0012

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.121390). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)
C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.110487). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 14s 165us/step - loss: 0.0013 ETA: 0s - loss: 0.001
Epoch 54/100000
87006/87006 [==============================] - 10s 118us/step - loss: 6.4220e-04
Epoch 55/100000
87006/87006 [==============================] - 11s 125us/step - loss: 7.1141e-040s - loss: 
Epoch 56/100000
 7800/87006 [=>............................] - ETA: 23s - loss: 6.5052e-04

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.104343). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 11s 122us/step - loss: 6.5802e-04
Epoch 57/100000
87006/87006 [==============================] - 11s 125us/step - loss: 5.8011e-042s - loss: 4.5479e - ETA: 2s - loss: 4.49 - ETA:
Epoch 58/100000
87006/87006 [==============================] - 8s 89us/step - loss: 0.0012: 9s - loss - ETA:
Epoch 59/100000
87006/87006 [==============================] - 10s 114us/step - loss: 5.0915e-040
Epoch 60/100000
87006/87006 [==============================] - 11s 130us/step - loss: 5.4722e-04
Epoch 61/100000
87006/87006 [==============================] - 9s 107us/step - loss: 4.6552e-04
Epoch 62/100000
87006/87006 [==============================] - 10s 111us/step - loss: 5.4190e-04- E
Epoch 63/100000
87006/87006 [==============================] - 6s 72us/step - loss: 5.1234e-04: 4s - lo - ETA: 2s - l - ETA: 0s - loss: 4.34
Epoch 64/100000
87006/87006 [==============================] - 10s 111us/step - loss: 7.0432e-04
Epoch 65/100000
87006/87006 [=======

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.162782). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)
C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.104836). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 7s 82us/step - loss: 4.0314e-04
Epoch 69/100000
87006/87006 [==============================] - 4s 42us/step - loss: 6.4188e-04
Epoch 70/100000
87006/87006 [==============================] - 4s 49us/step - loss: 4.6858e-04
Epoch 71/100000
87006/87006 [==============================] - 4s 41us/step - loss: 3.4738e-04
Epoch 72/100000
87006/87006 [==============================] - 5s 59us/step - loss: 5.1441e-04
Epoch 73/100000
87006/87006 [==============================] - 4s 49us/step - loss: 4.7704e-04
Epoch 74/100000
87006/87006 [==============================] - 7s 84us/step - loss: 5.6626e-04
Epoch 75/100000
87006/87006 [==============================] - 7s 77us/step - loss: 3.7360e-04:  - ETA: 5s - loss: 3.02
Epoch 76/100000
87006/87006 [==============================] - 8s 97us/step - loss: 3.7452e-04: 1s - loss: 3.0760e - E - ETA:  - ETA: 0s - lo - ETA: 0s - loss: 3.7540e-
Epoch 77/100000
79200/87006 [==========================>...] -

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.107986). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 12s 137us/step - loss: 4.2048e-04
Epoch 78/100000
87006/87006 [==============================] - 11s 126us/step - loss: 3.7430e-04
Epoch 79/100000
87006/87006 [==============================] - 8s 97us/step - loss: 5.2397e-04
Epoch 80/100000
87006/87006 [==============================] - 11s 125us/step - loss: 4.0213e-04
Epoch 81/100000
87006/87006 [==============================] - 14s 162us/step - loss: 3.5433e-041s - loss: 3.8 - ETA: 0s - loss: 
Epoch 82/100000
  900/87006 [..............................] - ETA: 22s - loss: 8.9690e-05

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.134154). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


 1500/87006 [..............................] - ETA: 34s - loss: 1.1607e-04

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.105458). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 13s 153us/step - loss: 4.1836e-044
Epoch 83/100000
87006/87006 [==============================] - 9s 98us/step - loss: 3.8217e-04
Epoch 84/100000
87006/87006 [==============================] - 5s 63us/step - loss: 4.4099e-04: 0s - loss: 4.
Epoch 85/100000
87006/87006 [==============================] - 13s 152us/step - loss: 4.0905e-04ETA: 8s - loss: 1.6576e - ETA: 8s - loss: 1 - ETA: - ETA: 6s -  - ETA: 2s - - ETA: 0s - loss: 4
Epoch 86/100000
87006/87006 [==============================] - 12s 133us/step - loss: 4.9428e-049s - loss: 2. - ETA: 9s - loss: 2.0761e - ETA: 9s - ETA: 0s - loss: 4.9606e-
Epoch 87/100000
87006/87006 [==============================] - 10s 120us/step - loss: 2.4214e-042s - loss: 2.6727e-0 - ETA: 2
Epoch 88/100000
87006/87006 [==============================] - 9s 107us/step - loss: 4.0772e-04 4s - los - ETA: 2s - loss: 4.87 - ETA: 2s - lo
Epoch 89/100000
87006/87006 [==============================] - 5s 62us/step - l

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.152267). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)
C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.139073). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


 3300/87006 [>.............................] - ETA: 18s - loss: 6.3547e-04

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.107221). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 12s 136us/step - loss: 3.4738e-040s - loss: 3.615 - ETA: 0s - loss: 3.
Epoch 115/100000
87006/87006 [==============================] - 11s 130us/step - loss: 2.1702e-045 - ETA: 4s - loss: 2.8189 - ETA: 3s - loss:  - ETA: 2s
Epoch 116/100000
87006/87006 [==============================] - 12s 138us/step - loss: 3.2703e-048s - loss: 2.7800e - ETA: 8s - loss: 2.7 - ETA: 0s - loss
Epoch 117/100000
87006/87006 [==============================] - 12s 139us/step - loss: 2.8877e-04
Epoch 118/100000
87006/87006 [==============================] - 10s 112us/step - loss: 2.7274e-04 ETA: 11s -  - ETA: 14s - loss: 6. - ETA: 15s - loss: 6.9001e- - E - ETA: 0s - loss: 2.
Epoch 119/100000
87006/87006 [==============================] - 8s 94us/step - loss: 2.6903e-04
Epoch 120/100000
87006/87006 [==============================] - 9s 107us/step - loss: 3.4503e-04 4s - loss: 6.572 - ETA: 5s - loss:  - ETA: 5s - loss - ETA: 4s - lo - ETA: 5s -
Epoch 121/100000
8

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.100287). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 10s 114us/step - loss: 2.6228e-04
Epoch 124/100000
87006/87006 [==============================] - 11s 126us/step - loss: 3.3858e-04
Epoch 125/100000
87006/87006 [==============================] - 13s 150us/step - loss: 3.6294e-04ETA: 1s
Epoch 126/100000
87006/87006 [==============================] - 8s 94us/step - loss: 2.1316e-04
Epoch 127/100000
87006/87006 [==============================] - 5s 53us/step - loss: 2.0305e-04: 0s - loss: 2.04
Epoch 128/100000
87006/87006 [==============================] - 10s 109us/step - loss: 3.2659e-044s - lo  - - ETA: 0s - loss: 3.2783e-
Epoch 129/100000
87006/87006 [==============================] - 5s 57us/step - loss: 2.3252e-04: 5s - los - ETA: 0s - loss: 2.
Epoch 130/100000
87006/87006 [==============================] - 12s 136us/step - loss: 2.1947e-044 - ETA
Epoch 131/100000
87006/87006 [==============================] - 9s 102us/step - loss: 2.8605e-04
Epoch 132/100000
87006/87006 [=============

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.119875). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 13s 153us/step - loss: 2.6952e-042s - loss: 2.751 - ET
Epoch 150/100000
87006/87006 [==============================] - 15s 167us/step - loss: 2.2592e-041s - los
Epoch 151/100000
 1200/87006 [..............................] - ETA: 23s - loss: 5.6752e-04

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.130975). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 12s 141us/step - loss: 1.7101e-04
Epoch 152/100000
87006/87006 [==============================] - 10s 114us/step - loss: 2.3707e-04
Epoch 153/100000
87006/87006 [==============================] - 11s 132us/step - loss: 2.5507e-04
Epoch 154/100000
87006/87006 [==============================] - 10s 113us/step - loss: 2.3412e-043s 
Epoch 155/100000
87006/87006 [==============================] - 14s 165us/step - loss: 3.1285e-04
Epoch 156/100000
87006/87006 [==============================] - 8s 95us/step - loss: 2.0733e-04
Epoch 157/100000
87006/87006 [==============================] - 11s 129us/step - loss: 1.2237e-041s - loss: 1.1947e-0 - ETA: 1s  - ETA: 0s - loss: 1.
Epoch 158/100000
87006/87006 [==============================] - 11s 128us/step - loss: 1.8311e-0410s - loss:  - ETA: 9 - ETA: 9 - 
Epoch 159/100000
87006/87006 [==============================] - 6s 66us/step - loss: 2.6886e-04
Epoch 160/100000
87006/87006 [=====================

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.101283). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 13s 155us/step - loss: 1.7813e-043s - loss: - ETA: 0s - loss: 1.
Epoch 178/100000
87006/87006 [==============================] - 10s 116us/step - loss: 1.6986e-043s  - ETA: 2s - loss: 1.7 - ETA: 2 - ETA - ETA: 0s - loss: 1.691
Epoch 179/100000
 1500/87006 [..............................] - ETA: 14s - loss: 7.4689e-05

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.108283). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 13s 145us/step - loss: 2.2476e-04 - ETA: 3s - loss: 2. - ETA: 3s - lo - E - ETA: 2s - los - ETA: 1s - loss: 2 - ETA: 1
Epoch 180/100000
87006/87006 [==============================] - 12s 137us/step - loss: 1.7697e-046s  - ETA: 6s - loss: 1.0197e-0 - ETA: 
Epoch 181/100000
87006/87006 [==============================] - 9s 100us/step - loss: 1.2794e-04
Epoch 182/100000
87006/87006 [==============================] - 6s 64us/step - loss: 1.6175e-04: 0s - loss: 1.6213e-
Epoch 183/100000
87006/87006 [==============================] - 13s 145us/step - loss: 1.4626e-040s - loss: 1.4820e - ETA: 0s - loss: 1.4832
Epoch 184/100000
87006/87006 [==============================] - 10s 120us/step - loss: 1.8265e-04
Epoch 185/100000
59400/87006 [===================>..........] - ETA: 5s - loss: 4.5254e-04- ETA: 8s - loss: 1.9492 - ET

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.103043). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


60300/87006 [===================>..........] - ETA: 5s - loss: 4.6107e-04

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.102258). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


61800/87006 [====================>.........] - ETA: 5s - loss: 4.6746e-04

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.102737). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)
C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.101361). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 17s 193us/step - loss: 3.6901e-044s
Epoch 186/100000
87006/87006 [==============================] - 10s 113us/step - loss: 1.4695e-04
Epoch 187/100000
87006/87006 [==============================] - 8s 96us/step - loss: 1.5401e-04
Epoch 188/100000
87006/87006 [==============================] - 11s 130us/step - loss: 2.4061e-04
Epoch 189/100000
87006/87006 [==============================] - 13s 150us/step - loss: 1.1634e-04
Epoch 190/100000
87006/87006 [==============================] - 13s 151us/step - loss: 1.5672e-044s - loss: 2.2404 - E - 
Epoch 191/100000
87006/87006 [==============================] - 13s 146us/step - loss: 1.6672e-045s - l - ETA: 
Epoch 192/100000
87006/87006 [==============================] - 6s 69us/step - loss: 1.3181e-04
Epoch 193/100000
87006/87006 [==============================] - 11s 127us/step - loss: 1.4474e-046s -  - ETA: 1s -
Epoch 194/100000
87006/87006 [==============================] - 6s 74us/step - los

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.168297). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 10s 115us/step - loss: 6.5940e-0513s - lo - ETA: 14s - loss: 2. - ETA: 12s  - ETA: 1
Epoch 238/100000
87006/87006 [==============================] - 15s 167us/step - loss: 1.2962e-045s - loss: 9.2979e-0 - ETA - ETA: 4s -  - ETA: 2s 
Epoch 239/100000
87006/87006 [==============================] - 8s 95us/step - loss: 5.6907e-05: 2s - loss: - ETA: 0s - loss: 5.6375
Epoch 240/100000
87006/87006 [==============================] - 7s 79us/step - loss: 9.0246e-05
Epoch 241/100000
87006/87006 [==============================] - 11s 123us/step - loss: 8.1631e-05
Epoch 242/100000
87006/87006 [==============================] - ETA: 0s - loss: 8.3588e-0 - 12s 137us/step - loss: 8.3588e-05
Epoch 243/100000
87006/87006 [==============================] - 11s 121us/step - loss: 8.0347e-05 - ETA: 0s - lo - ETA: 0s - loss: 8.0342e-0
Epoch 244/100000
87006/87006 [==============================] - 10s 119us/step - loss: 9.1653e-05
Epoch 245/100000
87006/87006

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.100278). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 15s 167us/step - loss: 7.0233e-0511s - ETA: 9s - loss:  - ETA: 9s - loss: - ETA: 8s - loss: 6.5847e-0 - ETA: 8s - loss: - ETA: 7s - loss: 6. - ETA: 7s - loss: 7. - ET - ETA - ETA: 1s - loss: 7.312 - ETA: 1s - loss: 7. - ETA: 0s - loss: 7 - ETA: 0s - loss:
Epoch 276/100000
87006/87006 [==============================] - 10s 119us/step - loss: 6.4565e-05
Epoch 277/100000
87006/87006 [==============================] - 13s 152us/step - loss: 4.5718e-05
Epoch 278/100000
 1200/87006 [..............................] - ETA: 33s - loss: 5.1261e-05 ETA: 21s - loss: 1.7498

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.150663). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 14s 158us/step - loss: 6.6588e-052s
Epoch 279/100000
87006/87006 [==============================] - 10s 115us/step - loss: 6.3743e-05
Epoch 280/100000
87006/87006 [==============================] - 5s 62us/step - loss: 4.1575e-05: 
Epoch 281/100000
87006/87006 [==============================] - 10s 115us/step - loss: 5.6275e-053s - loss: 7.03 - ETA: 3s - loss: - ETA: 1s - loss: 5.9
Epoch 282/100000
87006/87006 [==============================] - ETA: 0s - loss: 5.0893e-05- ETA: 2s - loss: 5.62  - 11s 130us/step - loss: 5.1083e-05
Epoch 283/100000
87006/87006 [==============================] - 7s 86us/step - loss: 8.1000e-05: 0s - loss: 8. - ETA: 0s - loss:
Epoch 284/100000
87006/87006 [==============================] - 8s 95us/step - loss: 4.2973e-05
Epoch 285/100000
87006/87006 [==============================] - 6s 66us/step - loss: 6.6539e-05
Epoch 286/100000
87006/87006 [==============================] - 6s 70us/step - loss: 6.2377e-05
E

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.109283). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 9s 101us/step - loss: 3.5660e-05 2s - loss -  - ETA: 0s - loss: 3.6
Epoch 292/100000
87006/87006 [==============================] - 10s 119us/step - loss: 5.9706e-051s - loss: 
Epoch 293/100000
87006/87006 [==============================] - 11s 126us/step - loss: 4.8462e-058s - loss: 5.1779e-05 - ETA: 9s - ETA: 9s - loss: 3.0367e-0 - ETA: 9s - loss: 3.0003e - ETA: 9s - loss: -  - ETA: 6s - loss: 3.8507 - ETA: 6s - loss: -  - ETA: 3 - E
Epoch 294/100000
87006/87006 [==============================] - 7s 78us/step - loss: 5.1622e-05
Epoch 295/100000
87006/87006 [==============================] - 8s 96us/step - loss: 4.3953e-05: 0s - loss: 4.5365 - ETA: 0s - loss: 4.5178 - ETA: 0s - loss: 
Epoch 296/100000
87006/87006 [==============================] - 9s 103us/step - loss: 4.8591e-05- 
Epoch 297/100000
87006/87006 [==============================] - 10s 115us/step - loss: 5.4893e-053s  - ETA: 3s - los - ETA: 3s - ETA: 0s - loss: 
Epoch 298/100

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.181824). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)
C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.153424). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


 3000/87006 [>.............................] - ETA: 14s - loss: 1.5270e-

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.125023). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 9s 101us/step - loss: 5.0384e-05 9 - 
Epoch 302/100000
87006/87006 [==============================] - 8s 94us/step - loss: 3.9762e-05: 1
Epoch 303/100000
87006/87006 [==============================] - 11s 130us/step - loss: 5.2802e-05
Epoch 304/100000
87006/87006 [==============================] - 8s 90us/step - loss: 4.2583e-05
Epoch 305/100000
87006/87006 [==============================] - 11s 124us/step - loss: 3.7890e-058s - loss: 3.9060e-0 - ETA: 8s - loss - ETA: 0s - lo
Epoch 306/100000
87006/87006 [==============================] - 10s 118us/step - loss: 4.5124e-05
Epoch 307/100000
87006/87006 [==============================] - 12s 133us/step - loss: 5.0943e-055s - los - ETA: 5s - loss: 6.32 - ETA: 4s - loss: 6.0 - ETA: 4s - loss: 5.7891 - ETA: 4s - loss
Epoch 308/100000
87006/87006 [==============================] - 9s 100us/step - loss: 3.9734e-05 1s - loss: 3 - ETA: 1s - loss: 3.8930 - ETA: 1s -
Epoch 309/100000
87006/87006 [====

C:\Users\User\.conda\envs\theano_v1\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.100279). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


87006/87006 [==============================] - 11s 122us/step - loss: 4.3749e-05TA: 0s - loss: 4
Epoch 323/100000
87006/87006 [==============================] - 14s 155us/step - loss: 3.5027e-050s - loss: 3.5606e-0 - ETA: 0s - loss: 3.55
Epoch 324/100000
87006/87006 [==============================] - 13s 154us/step - loss: 4.2460e-05
Epoch 325/100000
87006/87006 [==============================] - 14s 156us/step - loss: 4.7299e-054s - lo -
Epoch 326/100000
87006/87006 [==============================] - 11s 121us/step - loss: 2.8219e-05 5s - loss:  - ETA: 6s - loss:  - ETA: 5s - loss: - ETA - ETA: 1s - loss:  - ETA: 0s - loss:  - ETA: 0s - loss: 2.8
Epoch 327/100000
87006/87006 [==============================] - 12s 143us/step - loss: 4.6238e-05
Epoch 328/100000
87006/87006 [==============================] - 8s 97us/step - loss: 6.2157e-05
Epoch 329/100000
87006/87006 [==============================] - 13s 152us/step - loss: 2.1266e-058s  - - ETA: 6s - loss: 1.06 - ETA: 5s - loss: 1.0475

KeyboardInterrupt: 

In [20]:
### Now we load the test data and predict the model out of that
### What we do next is calculate R-squared and display it.
test_x = np.load('test_sample.npy')#,dtype=theano.config.floatX)
test_y= np.load('test_target.npy')#,dtype=theano.config.floatX)
test_x = np.load('test_sample.npy')#,dtype=theano.config.floatX)
test_y= np.load('test_target.npy')#,dtype=theano.config.floatX)
sample_out= model.predict(test_x)
ss_tot = np.sum((test_y-np.mean(test_y))**2)
SS_res = np.sum((sample_out.reshape(-1)-test_y)**2)
print("Rsquared:",(1 - SS_res/ss_tot)*100)

Rsquared: 99.95008688561789


R-squared of 99.95 is not great, but also not bad for first training. We can still improve this by retraining, just like we did it when we used lasagne package!

In [21]:
model.predict(test_x)

array([[1.505761 ],
       [1.5068358],
       [1.5069517],
       ...,
       [1.2960402],
       [1.2948159],
       [1.294792 ]], dtype=float32)

In [22]:
### And we save the model;

model.save("model.h1")